infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/SessionManager.java

```diff
@@ -125,32 +145,28 @@ public SessionManager(@Nonnull TopologyDescriptor topology, CorfuRuntime corfuRu
         this.metadataManager = new LogReplicationMetadataManager(corfuRuntime, topology.getTopologyConfigId());
         this.configManager = new LogReplicationConfigManager(runtime);
         this.upgradeManager = null;
-        replicationContext = new LogReplicationContext(configManager, topology.getTopologyConfigId(), localCorfuEndpoint);
-
-        createSessions();
+        this.replicationContext = new LogReplicationContext(configManager, topology.getTopologyConfigId(), localCorfuEndpoint);
+        this.replicationManager = replicationManager;
+        this.router = router;
     }
 
+
     /**
      * Refresh sessions based on new topoloogy
+     * The metadata updates will be done only by the leader, so other nodes do not overwrite the data
      *
      * @param newTopology   the new discovered topology
      */
-    // TODO [V2]: In the connectionModel PR, ensure that refresh is called by the leader node only. Otherwise the
-    //  metadata/status tables will be overwritten by all the nodes in the cluster
     public synchronized void refresh(@Nonnull TopologyDescriptor newTopology) {
```

Comment:
refresh() gets invoked from 1)onTopologyChange() and 2)lockAcquire().  In the first case, it is invoked on all nodes regardless of the leadership.

We should keep the behavior consistent.  Also, I think the in-memory sessions should be updated on all nodes.  Metadata tables can be updated only by the leader node.

Reply:
reiterating to keep the context: we do have the in-memory sessions updated on all the nodes.

The correct place would be to have refresh being called from onLeadershipAcquire and from onTopologyChange...
reason: we want the leader to write to the system tables, so we need to call refresh after leadership has been acquired.
And since we want the in-memory sessions to be on all the nodes, we call from onTopologyChange as well.

calling refresh from onLeadershipAcquire would lead to refresh being called twice unnecessarily. So to avoid that, I had one in lockAcquire instead of onLeadershipAcquire.